crates/bevy_ecs/src/schedule/system_graph.rs

```diff
@@ -0,0 +1,472 @@
+/// # use bevy_ecs::prelude::*;
+/// # fn sys_a() {}
+/// # fn sys_b() {}
+/// # fn sys_c() {}
+/// # fn sys_d() {}
+/// # fn sys_e() {}
+/// let graph = SystemGraph::new();
+///
+/// let start_a = graph.root(sys_a);
+/// let start_b = graph.root(sys_b);
+/// let start_c = graph.root(sys_c);
+///
+/// (start_a, start_b, start_c)
+///     .join(sys_d)
+///     .then(sys_e);
+///
+/// // Convert into a SystemSet
+/// let system_set: SystemSet = graph.into();
+/// ```
+///
+/// # Fan Out into Fan In
+/// The types used to implement [fork] and [join] are composable.
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # fn sys_a() {}
+/// # fn sys_b() {}
+/// # fn sys_c() {}
+/// # fn sys_d() {}
+/// # fn sys_e() {}
+/// # fn sys_f() {}
+/// let graph = SystemGraph::new();
+/// graph.root(sys_a)
+///      .fork((sys_b, sys_c, sys_d))
+///      .join(sys_e)
+///      .then(sys_f);
+///
+/// // Convert into a SystemSet
+/// let system_set: SystemSet = graph.into();
+/// ```
+///
+/// # Cloning
+/// This type is backed by a Rc, so cloning it will still point to the same logical
+/// underlying graph.
+///
+/// [fork]: crate::schedule::SystemGraphNode::fork
+/// [join]: crate::schedule::SystemJoin::join
+#[derive(Clone)]
+pub struct SystemGraph {
+    id: u32,
+    nodes: Rc<RefCell<HashMap<NodeId, SystemDescriptor>>>,
+}
+
+impl Default for SystemGraph {
+    fn default() -> Self {
+        Self {
+            id: NEXT_GRAPH_ID.fetch_add(1, Ordering::Relaxed),
+            nodes: Default::default(),
+        }
+    }
+}
+
+impl SystemGraph {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[cfg(test)]
+    pub(crate) fn with_id(id: u32) -> Self {
+        Self {
+            id,
+            nodes: Default::default(),
+        }
+    }
+
+    /// Creates a root graph node without any dependencies. A graph can have multiple distinct
+    /// root nodes.
+    pub fn root<Params>(&self, system: impl IntoSystemDescriptor<Params>) -> SystemGraphNode {
+        self.create_node(system.into_descriptor())
+    }
+
+    /// Checks if two graphs instances point to the same logical underlying graph.
+    pub fn is_same_graph(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+
+    fn create_node(&self, mut system: SystemDescriptor) -> SystemGraphNode {
+        let mut nodes = self.nodes.borrow_mut();
+        assert!(
+            nodes.len() <= u32::MAX as usize,
+            "Cannot add more than {} systems to a SystemGraph",
+            u32::MAX
+        );
+        let id = NodeId(self.id, nodes.len() as u32);
+        match &mut system {
+            SystemDescriptor::Parallel(descriptor) => descriptor.labels.push(id.dyn_clone()),
+            SystemDescriptor::Exclusive(descriptor) => descriptor.labels.push(id.dyn_clone()),
+        }
+        nodes.insert(id, system);
+        SystemGraphNode {
+            id,
+            graph: self.clone(),
+        }
+    }
+
+    fn add_dependency(&self, src: NodeId, dst: NodeId) {
```

Comment:
`dst` is not a clear variable name in context; I can't even guess what this is :)